x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/InitialSecurityConfigurationListener.java

```diff
@@ -0,0 +1,191 @@
+import org.elasticsearch.xpack.security.support.SecurityIndexManager;
+
+import java.io.PrintStream;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+import static org.elasticsearch.xpack.security.authc.esnative.ReservedRealm.BOOTSTRAP_ELASTIC_PASSWORD;
+import static org.elasticsearch.xpack.security.tool.CommandUtils.generatePassword;
+
+public class InitialSecurityConfigurationListener implements BiConsumer<SecurityIndexManager.State, SecurityIndexManager.State> {
+
+    private static final String tokenKey = "token";
+    private static final String fingerprintKey = "caCertFingerprint";
+    private static final String passwordKey = "elasticPassword";
+
+    private static final Logger LOGGER = LogManager.getLogger(InitialSecurityConfigurationListener.class);
+    private final NativeUsersStore nativeUsersStore;
+    private final SecurityIndexManager securityIndexManager;
+    private final Environment environment;
+    private final SSLService sslService;
+    private final Client client;
+
+    public InitialSecurityConfigurationListener(
+        NativeUsersStore nativeUsersStore,
+        SecurityIndexManager securityIndexManager,
+        SSLService sslService,
+        Client client,
+        Environment environment
+    ) {
+        this.nativeUsersStore = nativeUsersStore;
+        this.securityIndexManager = securityIndexManager;
+        this.sslService = sslService;
+        this.client = client;
+        this.environment = environment;
+    }
+
+    @Override
+    public void accept(SecurityIndexManager.State previousState, SecurityIndexManager.State currentState) {
+        final PrintStream out = BootstrapInfo.getOriginalStandardOut();
+        // Check if it has been closed, try to write something so that we trigger PrintStream#ensureOpen
+        out.println();
+        if (out.checkError()) {
+            outputOnError(null);
+            return;
+        }
+        if (previousState.equals(SecurityIndexManager.State.UNRECOVERED_STATE)
+            && currentState.equals(SecurityIndexManager.State.UNRECOVERED_STATE) == false
+            && securityIndexManager.indexExists() == false
+            && XPackSettings.ENROLLMENT_ENABLED.get(environment.settings())) {
+            GroupedActionListener<Map<String, String>> groupedActionListener = new GroupedActionListener<>(ActionListener.wrap(results -> {
+                final Map<String, String> allResultsMap = new HashMap<>();
+                for (Map<String, String> result : results) {
+                    allResultsMap.putAll(result);
+                }
+                final String password = allResultsMap.get(passwordKey);
+                final String token = allResultsMap.get(tokenKey);
+                final String caCertFingerprint = allResultsMap.get(fingerprintKey);
+                outputInformationToConsole(password, token, caCertFingerprint, out);
```

Comment:
These all can be null but we cater for this when printing the information to stdout in `outputInformationToConsole`